### PR TITLE
Filter Module Cache for Coin  

### PR DESCRIPTION
**Pull request by emilypi**
This PR filters the module cache for just coin after the next service date.

### CODE REVIEW

File: src/Chainweb/Version.hs [hunk 1]
```diff
@@ -1007,6 +1008,21 @@ chainweb216Pact aoa v h = case aoa of
...
+    :: AtOrAfter
+    -> ChainwebVersion
+    -> BlockHeight
+    -> Bool
+chainweb217Pact aoa v h = case aoa of
+    At -> go (==) v h
+    After -> go (<) v h
+  where
+    go f Mainnet01 = f 3_250_357 -- 2022-12-01 00:00:00+00:00
+    go f Testnet04 = f 2_777_376 -- 2022-12-01 12:00:00+00:00
```
[the remark this answers is not in the document]
Comment (larskuhtz) [reply src/Chainweb/Version.hs:1021]: That's what I got:

```
|    | network   | time                      |   height |
|---:|:----------|:--------------------------|---------:|
|  0 | testnet   | 2022-12-01 12:00:00+00:00 |  2777370 |
|  1 | mainnet   | 2022-12-02 00:00:00+00:00 |  3250353 |
```

So, I think we are good.

File: src/Chainweb/Pact/TransactionExec.hs [hunk 2]
```diff
@@ -517,6 +544,7 @@ applyUpgrades v cid height
      | pact4coin3Upgrade At v height = applyCoinV3
      | chainweb214Pact At v height = applyCoinV4
      | chainweb215Pact At v height = applyCoinV5
+     | chainweb217Pact At v height = filterModuleCache
```
Comment (larskuhtz) [inline src/Chainweb/Pact/TransactionExec.hs:547]: @emilypi @ggobugi27 please make sure that there is a regression test for this

File: src/Chainweb/Pact/Types.hs [hunk 3]
```diff
@@ -449,10 +449,17 @@ updateInitCache :: ModuleCache -> PactServiceM cas ()
...
+        ph = _parentHeader _psParentHeader
+        pbh = bf $ _blockHeight ph
+
+    v <- view psVersion
+
     psInitCache .= case M.lookupLE pbh _psInitCache of
       Nothing -> M.singleton pbh mc
-      Just (_,before) -> M.insert pbh (HM.union mc before) _psInitCache
+      Just (_,before)
+        | chainweb217Pact After v pbh || chainweb217Pact At v pbh->
```
Comment (larskuhtz) [inline src/Chainweb/Pact/Types.hs:460]:
```suggestion
        | chainweb217Pact After v pbh || chainweb217Pact At v pbh ->
```

File: test/Chainweb/Test/Pact/ModuleCacheOnRestart.hs [hunk 4]
```diff
@@ -35,6 +35,7 @@ import Pact.Types.Term
 
 -- chainweb imports
 
+
```
Comment (larskuhtz) [inline test/Chainweb/Test/Pact/ModuleCacheOnRestart.hs:38]:
```suggestion
```

**Review by larskuhtz — APPROVED** [review]
just a couple minor suggestions

**Comment by larskuhtz** [comment]
There is a flake test failure in one of the Mac builds. Not sure if it is related to this PR, though.

```
Chainweb Tests
  Chainweb-Pact Tests
    Chainweb.Test.Pact.PactReplay
      on-restart:                                                                                            FAIL (5.34s)
        max block has height 0
        mine block on top of max block (5.10s)
          PactReplay.mineBlock: Test timeout. Most likely a test case caused a pact service failure that wasn't caught, and the test was blocked while waiting for the result
          CallStack (from HasCallStack):
            error, called at test/Chainweb/Test/Pact/PactReplay.hs:312:16 in main:Chainweb.Test.Pact.PactReplay
        
        Use -p '/on-restart/' to rerun this test only.
```

File: src/Chainweb/Pact/Types.hs [hunk 5]
```diff
@@ -449,10 +449,17 @@ updateInitCache :: ModuleCache -> PactServiceM cas ()
 updateInitCache mc = get >>= \PactServiceState{..} -> do
     let bf 0 = 0
         bf h = succ h
-        pbh = bf . _blockHeight . _parentHeader $ _psParentHeader
+        ph = _parentHeader _psParentHeader
```
[the remark this answers is not in the document]
Comment (emilypi) [reply src/Chainweb/Pact/Types.hs:452]: Ah good eye - that's actually some stale work that should be reverted.

**Comment by emilypi** [comment]
@larskuhtz considering there was no substantive code changes between aa1dfce and e743efb, I don't think it's this commit in particular. We seem to have a flaky test.

**Review by jmcardon — APPROVED** [review]

**Pull request merged**